File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode prevGroup = dummy;
        
        while (true) {
            ListNode kth = getKth(prevGroup, k);
            if (kth == null) break;
            
            ListNode nextGroup = kth.next;
            ListNode[] reversed = reverse(prevGroup.next, kth);
            
            prevGroup.next = reversed[0];
            reversed[1].next = nextGroup;
            prevGroup = reversed[1];
        }
        
        return dummy.next;
    }
    
    private ListNode getKth(ListNode node, int k) {
        while (node != null && k > 0) {
            node = node.next;
            k--;
        }
        return node;
    }
    
    private ListNode[] reverse(ListNode head, ListNode tail) {
        ListNode prev = tail.next;
        ListNode curr = head;
        while (prev != tail) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return new ListNode[]{tail, head};
    }
}

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: Optional[ListNode]
        :type k: int
        :rtype: Optional[ListNode]
        """
        # Check if there are k nodes to reverse
        curr = head
        for i in range(k):
            if not curr:
                return head
            curr = curr.next
        
        # Reverse first k nodes
        prev = None
        curr = head
        for i in range(k):
            next_temp = curr.next
            curr.next = prev
            prev = curr
            curr = next_temp
        
        # Recursively reverse remaining groups
        head.next = self.reverseKGroup(curr, k)
        
        return prev